### PR TITLE
[LBSE] Fix clipping + SVGUseElement support

### PR DESCRIPTION
**Pull request by nikolaszimmermann**
#### e94cf196874e3b91550e8f4a3bf5c44aa318798a
<pre>
[LBSE] Fix clipping + SVGUseElement support
<a href="https://bugs.webkit.org/show_bug.cgi?id=265244">https://bugs.webkit.org/show_bug.cgi?id=265244</a>

Reviewed by Rob Buis.

Teach pathFromGraphicsElement() about SVGUseElement, and use that
to support &lt;use&gt; as &lt;clipPath&gt; children in LBSE. The legacy SVG
engine uses the deprecated virtual SVGUseElement::toClipPath()
function -- LBSE no longer makes use of these code paths.

Covered by existing tests - 5+ fixed tests in LBSE.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/svg/SVGPathData.cpp:
(WebCore::pathFromUseElement):
(WebCore::pathFromGraphicsElement):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::shouldApplyPathClipping const):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::toClipPath):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::toClipPath):
(WebCore::SVGUseElement::clipChild const):
* Source/WebCore/svg/SVGUseElement.h:

Canonical link: <a href="https://commits.webkit.org/271101@main">https://commits.webkit.org/271101@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/348f6166b0469fd914dad26b676641e9faf73457

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27274 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5913 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29497 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24956 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27742 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24780 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4711 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23407 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4106 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4224 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30136 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24820 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30395 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28338 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5725 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6584 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4718 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4639 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/svg/SVGPathData.cpp [hunk 1]
```diff
@@ -183,6 +184,25 @@ static Path pathFromRectElement(const SVGElement& element)
     return path;
 }
 
+static Path pathFromUseElement(const SVGElement& element)
+{
+    const auto& useElement = downcast<SVGUseElement>(element);
```
Comment (rwlbuis) [inline Source/WebCore/rendering/svg/SVGPathData.cpp:189]: We probably want to assert the downcast is safe.

File: Source/WebCore/svg/SVGUseElement.cpp [hunk 2]
```diff
@@ -338,6 +341,16 @@ Path SVGUseElement::toClipPath()
     return path;
 }
 
+RefPtr<SVGElement> SVGUseElement::clipChild() const
+{
+    auto targetClone = this->targetClone();
+    if (!targetClone)
+        return nullptr;
+    if (!isDirectReference(*targetClone))
+        return nullptr;
```
Comment (rwlbuis) [inline Source/WebCore/svg/SVGUseElement.cpp:348]: Style nit, the two ifs can be combined into one.

File: Source/WebCore/svg/SVGUseElement.cpp [hunk 3]
```diff
@@ -338,6 +341,16 @@ Path SVGUseElement::toClipPath()
...
+    return targetClone.get();
```
Comment (rwlbuis) [inline Source/WebCore/svg/SVGUseElement.cpp]: Probably get() is not needed here?

**Review by rwlbuis — APPROVED** [review]
LGTM.

File: Source/WebCore/rendering/svg/SVGPathData.cpp (hunk 1 above)
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/SVGPathData.cpp:189]: That's already done by the callers -- see the other methods, they work the same.

File: Source/WebCore/svg/SVGUseElement.cpp (hunk 3 above)
Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGUseElement.cpp]: Correct, will fix!

File: Source/WebCore/svg/SVGUseElement.cpp (hunk 2 above)
Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGUseElement.cpp:348]: Ok, fixed.

**Pull request merged**